### PR TITLE
Allow custom prediction collectors

### PR DESCRIPTION
**Pull request by dfederm**
This is a bit of a redesign to support custom prediction collection.

This will enable high-performance scenario which don't want to translate the prediction object model to the caller's object model, reducing a ton of allocations.

As part of this I've refactored the `BuildItem` and subclasses into a single `PredictedItem` class. The "is directory" notion on inputs is now just inferred from the property it's obtained from (`InputFiles` vs `InputDirectories`).

### CODE REVIEW

File: src/BuildPrediction/DefaultProjectPredictionCollector.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        public void AddOutputDirectory(string path, string projectDirectory, string predictorName) => AddBuildItem(_outputDirectoriesByPath, path, projectDirectory, predictorName);
+
+        private static void AddBuildItem(Dictionary<string, PredictedItem> items, string path, string projectDirectory, string predictorName)
+        {
+            // Make the path absolute if needed.
+            if (!Path.IsPathRooted(path))
+            {
+                path = Path.GetFullPath(Path.Combine(projectDirectory, path));
+            }
+
```
Comment (dfederm) [inline src/BuildPrediction/DefaultProjectPredictionCollector.cs]: Note that path normalization is expected to happen in the collector. Will make sure this is (eventually) documented.

File: src/BuildPrediction/ProjectPredictionReporter.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Build.Prediction
+{
+    /// <summary>
+    /// A reporter <see cref="IProjectPredictor"/> instances use to report predictions.
```
Comment (erikma) [inline src/BuildPrediction/ProjectPredictionReporter.cs]: >  [](start = 18, length = 1)

that

File: src/BuildPrediction/ProjectPredictionReporter.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        /// <summary>
+        /// Report a prediction for an input file.
+        /// </summary>
+        /// <param name="path">The path of the file input.</param>
+        public void ReportInputFile(string path) => _predictionCollector.AddInputFile(path, _projectDirectory, _predictorName);
+
+        /// <summary>
+        /// Report a prediction for an input directory.
```
Comment (erikma) [inline src/BuildPrediction/ProjectPredictionReporter.cs]: >. [](start = 54, length = 1)

Implicitly this means the directory's contents, but not its subdirectories, are used as inputs. This is equivalent to a "*" glob.

File: src/BuildPrediction/Predictors/CSharpCompileItems.cs [hunk 4]
```diff
@@ -17,25 +14,16 @@ public class CSharpCompileItems : IProjectPredictor
...
+            ProjectPredictionReporter predictionReporter)
         {
-            // TODO: Need to determine how to normalize evaluated include selected below and determine if it is relative to project.
-            List<BuildInput> itemInputs = project.GetItems(CompileItemName)
-                .Select(item => new BuildInput(
-                    Path.Combine(project.DirectoryPath, item.EvaluatedInclude),
-                    isDirectory: false))
-                .ToList();
-            if (itemInputs.Count > 0)
+            var items = project.GetItems(CompileItemName);
```
Comment (erikma) [inline src/BuildPrediction/Predictors/CSharpCompileItems.cs]: >var [](start = 12, length = 3)

ugh. var.

**Pull request merged**